### PR TITLE
Closes #1436: Add support for html input type date for GeckoEngine

### LINKED ISSUE #1436: Add support for html input type date for GeckoEngine
User Story: https://github.com/mozilla-mobile/reference-browser/issues/223

### PR DESCRIPTION
**Pull request by Amejia481**
Depends on: https://github.com/mozilla-mobile/android-components/pull/1501

### CODE REVIEW

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/DatePickerDialogFragment.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    companion object {
+        /**
+         * A builder method for creating a [DatePickerDialogFragment]
+         * @param sessionId to create the dialog.
+         * @param title of the dialog.
+         * @param initialDate date that will be selected by default.
+         * @param minDate the minimumDate date that will be allowed to be selected.
+         * @param maxDate the maximumDate date that will be allowed to be selected.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/DatePickerDialogFragment.kt]: `@return` would be nice ;)

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 2]
```diff
@@ -161,6 +168,37 @@ class PromptFeature(
...
+                is PromptRequest.Date -> it.onSelect(selectedDate)
+            }
+            true
+        }
+    }
+
+    /**
+     * Event that is called when the user is requesting to clear the selected value from the dialog.
+     * This consumes the [PromptFeature] value from the [Session] indicated by [sessionId].
+     * @property sessionId that requested to show the dialog.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: `@param`

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt [hunk 3]
```diff
@@ -36,3 +39,17 @@ fun String.isPhone(): Boolean = contains("tel:", true)
 fun String.isEmail(): Boolean = contains("mailto:", true)
 
 fun String.isGeoLocation(): Boolean = contains("geo:", true)
+
+/**
+ * Converts a [String] to a [Date] object, if an empty string is given will return Date().
```
Comment (csadilek) [inline components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt]: Should add the second part to `@return ...., current date if empty string was provided`

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt [hunk 4]
```diff
@@ -36,3 +39,17 @@ fun String.isPhone(): Boolean = contains("tel:", true)
...
+ * @property format of the this given [String] object.
+ * @property locale in which you want to format this [String] object.
```
Comment (csadilek) [inline components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt]: `@param` ... locale the locale to use when converting the String, defaults to ....

File: docs/changelog.md [hunk 5]
```diff
@@ -30,7 +30,15 @@ permalink: /changelog/
...
 * **feature-prompts**
-  * Added support alerts dialogs.
+  * Added support for alerts dialogs.
+  * Added support for date picker dialogs.
+
+* **support-ktx**
+  New extension function `toDate` that converts a string to a Date object from a formatter input.
+  ```Kotlin
+       val date1 = "2019-11-28".toDate("yyyy-MM-dd")
+       val date1 = "".toDate("yyyy-MM-dd")) // Returns Date()
```
Comment (csadilek) [inline docs/changelog.md]: Not sure we need to show the second case as a feature i.e. when would you ever do this? :)

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt [hunk 6]
```diff
@@ -4,6 +4,8 @@
 
 package mozilla.components.concept.engine.prompt
 
+typealias JavaDate = java.util.Date
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt]: As we have `PromptRequest.Date`, the compiler gets confused about which `Date` class we're talking about.

  ```kotlin
sealed class PromptRequest {
  data class Date(
        val title: String,
        val initialDate: JavaDate,
        val minimumDate: JavaDate?,
        val maximumDate: JavaDate?,
        val onSelect: (JavaDate) -> Unit,
        val onClear: () -> Unit
    ) : PromptRequest()
}
```
https://github.com/mozilla-mobile/android-components/blob/39ddb959b57dcfec43585327d28e3014a57a7327/components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt#L62

I can also imported it directly?
  ```kotlin
sealed class PromptRequest {
  data class Date(
        val title: String,
        val initialDate: java.util.Date,
        val minimumDate: java.util.Date?,
        val maximumDate: java.util.Date?,
        val onSelect: (java.util.Date) -> Unit,
        val onClear: () -> Unit
    ) : PromptRequest()
}
```

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/prompt/GeckoPromptDelegate.kt [hunk 7]
```diff
@@ -100,6 +110,60 @@ internal class GeckoPromptDelegate(private val geckoEngineSession: GeckoEngineSe
...
+                    title ?: "",
+                    initialDateString,
+                    minDate,
+                    maxDate,
+                    onClear,
+                    "yyyy-'W'ww",
+                    geckoCallback
+                )
+            }
+            DATETIME_TYPE_TIME -> {
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/prompt/GeckoPromptDelegate.kt]: I'll include the issue #number, thanks for the catching it. 👍

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/DatePickerDialogFragment.kt (hunk 1 above)
Comment (Amejia481) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/DatePickerDialogFragment.kt]: :)

**Comment by Amejia481** [comment]
I updated the PR @csadilek,  
Thanks for the review, and being patient.
Sorry for all the grammar and docs nits 🙃 , I'll be more careful future PRs!

**Pull request merged**